tests/scripts/thread-cert/test_srp_auto_host_address.py

```diff
@@ -156,13 +156,14 @@ def test(self):
         self.check_registered_addresses(client, server)
 
         #-------------------------------------------------------------------
-        # Remove the on-mesh prefix (which will trigger an address to be
-        # removed) and check that the SRP client re-registered and updated
-        # server with the remaining address.
+        # Remove the on-mesh prefix  and wait for deprecating time and check
+        # that the SRP client re-registered and updated server with the
+        # remaining address.
 
         client.remove_prefix('fd00:abba:cafe:bee::/64')
         client.register_netdata()
-        self.simulator.go(5)
+
+        self.simulator.go(310)
```

Comment:
Why is this time increased here? If an address becomes deprecated, the SRP client should be reasonably quick about re-registering services using another (preferred) IPv6 address if that is available. And not wait until 300 sec when the deprecated address becomes invalid, that would be far too late.

Reply:
The current SRP client auto-address mode implementation will add all addresses (independent of their flags) excluding mesh-local/link-local. So we need to wait for it to deprecate.

It may be useful to update this (in future PR) on `Srp::Client`, i.e. add only the preferred addresses and include non-preferred one if there is no preferred address. Thoughts?

-----

There is another aspect related to this that is not ideal:
- In the code today when we update a `netif` address, we remove it and re-add it with new flags.
- This is signaled by `Notifier` as `kEventIp6AddressRemoved` and `kEventIp6AddressAdded` to all other modules including `Srp::Client`
- `Srp::Client` will then assume something has changed and tries to re-register.
- With this model, we may end up re-registering even when there is no change (only flags get updated).
- I think this may not a very common situation but I'd like us to improve this if we can at some point. 
   - The easy way I can see is to have `Srp::Client` track the set of addresses it registered last, though the downside is the increased memory/RAM requirement to track this.

Reply:
> It may be useful to update this (in future PR) on Srp::Client, i.e. add only the preferred addresses and include non-preferred one if there is no preferred address. Thoughts?

This makes sense to me. The purpose of preferred vs. deprecated states is to encourage new connections to use the preferred address, but allow existing connections using deprecated addresses to continue. Presumably, any client looking to resolve a AAAA record is also looking to establish a new connection. If SRP publishes both preferred and deprecated, there is no way to determine whether an address is in the preferred or deprecated state.

Reply:
The spec 14.3.5.1 has some requirements on this. Currently, it says deprecated addresses are never registered in SRP Updates. And as soon as a previously registered address changes its state (becomes deprecated, or becomes invalid) the client is required to reregister, without the deprecated/invalid addresses.   The ML-EID is always preferred so it is chosen if no preferred global or ULA address is available. That seems the best option to me.

Reply:
Submitted https://github.com/openthread/openthread/pull/9642 to update the "auto-address" mode to always exclude deprecating (non-preferred) addresses.

Reply:
> * With this model, we may end up re-registering even when there is no change (only flags get updated).
> * I think this may not a very common situation but I'd like us to improve this if we can at some point.
>   
>   * The easy way I can see is to have `Srp::Client` track the set of addresses it registered last, though the downside is the increased memory/RAM requirement to track this.

Submitted https://github.com/openthread/openthread/pull/9652 which adds mechnaism for SRP client to track which addresses it previously registered and only send update if there is a change to set of addresses it need to include in its SRP update registeration. The new mehcanim uses new bit-field in `Ip6::Netif::UnicastAddress` so should not add RAM/memory overhead.

Reply:
@abtink With #9642 and #9652 already merged, the wait time 310 should now be set back to 5 (or 10) seconds, I think?  Assuming this PR is / can be rebased on the latest main branch.